pkg/controller/group/group_controller.go

```diff
@@ -0,0 +1,131 @@
+		CreateFunc: func(createEvent event.CreateEvent) bool {
+			return createEvent.Meta.GetName() == clusterAdminGroupName
+		},
+		DeleteFunc: func(deleteEvent event.DeleteEvent) bool {
+			return deleteEvent.Meta.GetName() == clusterAdminGroupName
+		},
+		UpdateFunc: func(updateEvent event.UpdateEvent) bool {
+			return updateEvent.MetaNew.GetName() == clusterAdminGroupName
+		},
+		GenericFunc: func(genericEvent event.GenericEvent) bool {
+			return genericEvent.Meta.GetName() == clusterAdminGroupName
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileGroup implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileGroup{}
+
+// ReconcileGroup reconciles a Group object
+type ReconcileGroup struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client            client.Client
+	scheme            *runtime.Scheme
+	metricsAggregator *metrics.AdoptionMetricsAggregator
+}
+
+// Reconcile reads that state of the cluster for a Group object and makes changes based on the state read
+// and what is in the Group.Spec
+// TODO(user): Modify this Reconcile function to implement your Controller logic.  This example creates
+// a Pod as an example
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileGroup) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling Group")
+
+	// Fetch the Group group
+	group := &userv1.Group{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, group)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+	if group.ObjectMeta.DeletionTimestamp.IsZero() {
+		if !utils.ContainsString(group.Finalizers, finalizer) {
+			controllerutil.AddFinalizer(group, finalizer)
+			if err := r.client.Update(context.Background(), group); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+		if len(group.Users) > 0 {
+			r.metricsAggregator.SetClusterAdmin(true)
+		}
+	} else {
+		if utils.ContainsString(group.Finalizers, finalizer) {
+			controllerutil.RemoveFinalizer(group, finalizer)
+			if err := r.client.Update(context.Background(), group); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+		r.metricsAggregator.SetClusterAdmin(false)
```

Comment:
Consider reversing these, as it's possible the Update could fail but the finalizer is still removed. And either way the intent to delete the group means we want to switch the metric off, yeah?